rust/simd/mat_mat_mul/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+    use std::arch::wasm32::*;
+    // `Clear` op.
+    let mut ab0 = f32x4(0_f32, 0_f32, 0_f32, 0_f32);
+    let mut ab1 = f32x4(0_f32, 0_f32, 0_f32, 0_f32);
+    let mut ab2 = f32x4(0_f32, 0_f32, 0_f32, 0_f32);
+    let mut ab3 = f32x4(0_f32, 0_f32, 0_f32, 0_f32);
+    let mut ab = [[0.0; 4]; 4];
+
+    // `AddMatMul` op.
+    unsafe {
+        let a = pa as *const f32;
+        let b = pb as *const v128;
+        for i in 0..k {
+            let a = std::slice::from_raw_parts(a.offset(4 * i as isize), 4);
+            let b = v128_load(b.offset(i as isize));
+            ab0 = f32x4_add(ab0, f32x4_mul(f32x4_splat(a[0]), b));
+            ab1 = f32x4_add(ab1, f32x4_mul(f32x4_splat(a[1]), b));
+            ab2 = f32x4_add(ab2, f32x4_mul(f32x4_splat(a[2]), b));
+            ab3 = f32x4_add(ab3, f32x4_mul(f32x4_splat(a[3]), b));
+        }
+    }
+
+    // `Store` op.
+    unsafe {
+        v128_store(ab[0].as_mut_ptr() as *mut v128, ab0);
+        v128_store(ab[1].as_mut_ptr() as *mut v128, ab1);
+        v128_store(ab[2].as_mut_ptr() as *mut v128, ab2);
+        v128_store(ab[3].as_mut_ptr() as *mut v128, ab3);
+    }
+
+    ab
+}
+
+/// Formats thousands for the specified `u64` integer (helper function).
+fn fmt(n: u64) -> String {
+    n.to_string()
+        .as_bytes()
+        .rchunks(3)
+        .rev()
+        .map(std::str::from_utf8)
+        .collect::<Result<Vec<&str>, _>>()
+        .unwrap()
+        .join("_")
+}
+
+////////////////////////////////////////////////////////////////////////
+// Floating-point `f32` multiplications.
+
+/// Returns the number of instructions used for a loop performing
+/// element-wise multiplication of `224x4` tiles of matrices `A` and `B`
+/// using generic Rust operators.
+#[ic_cdk_macros::query]
+fn generic_f32() -> u64 {
+    let a = mats::M_4X4_F32.repeat(K / 4);
+    let b = mats::M_4X4_F32.repeat(K / 4);
+    let pa = a.as_ptr() as *const u8;
+    let pb = b.as_ptr() as *const u8;
+
+    for _ in 0..ITERATIONS {
+        let res = add_mat_mul_4x4_generic::<f32>(K, pa, pb);
+        std::hint::black_box(res);
+    }
+    // Assert the generic and SIMD matrix multiplication results are the same.
+    assert_eq!(
+        add_mat_mul_4x4_generic::<f32>(K, pa, pb),
+        add_mat_mul_4x4_simd_f32(K, pa, pb)
+    );
+
+    let counter = ic_cdk::api::performance_counter(0);
+    ic_cdk::println!(
+        "Generic WebAssembly (f32): {:>12} Wasm instructions",
+        fmt(counter)
+    );
+    counter
+}
+
+/// Returns the number of instructions used for a loop performing
+/// element-wise multiplication of `224x4` tiles of matrices `A` and `B`
+/// using loop auto-vectorization for generic Rust operators.
+///
+/// The following line enables auto-vectorization using WebAssembly SIMD instructions.
+#[target_feature(enable = "simd128")]
+#[ic_cdk_macros::query]
+fn auto_vectorization_f32() -> u64 {
+    let a = mats::M_4X4_F32.repeat(K / 4);
+    let b = mats::M_4X4_F32.repeat(K / 4);
```

Comment:
I was expecting that the number of rows in matrix A matches the number of columns in matrix B in order for multiplication to work.

It seems that we are multiplying matrix A with a transposed matrix B (rows and columns swapped?). If so, then let's add a comment about it to avoid confusing readers?